src/handlers/car-block.js

```diff
@@ -0,0 +1,82 @@
+/* eslint-env browser */
+/* global FixedLengthStream */
+import { HttpError } from '@web3-storage/gateway-lib/util'
+import { CAR_CODE } from '../constants.js'
+import * as Http from '../lib/http.js'
+
+/** @typedef {import('@web3-storage/gateway-lib').IpfsUrlContext} CarBlockHandlerContext */
+
+/**
+ * Handler that serves CAR files directly from R2.
+ *
+ * @type {import('@web3-storage/gateway-lib').Handler<CarBlockHandlerContext, import('../bindings').Environment>}
+ */
+export async function handleCarBlock (request, env, ctx) {
+  const { searchParams, dataCid } = ctx
+  if (!dataCid) throw new Error('missing data CID')
+  if (!searchParams) throw new Error('missing URL search params')
+
+  if (request.method !== 'HEAD' && request.method !== 'GET') {
+    throw new HttpError('method not allowed', { status: 405 })
+  }
+  if (dataCid.code !== CAR_CODE) {
+    throw new HttpError('not a CAR CID', { status: 400 })
+  }
+
+  const etag = `"${dataCid}"`
+  if (request.headers.get('If-None-Match') === etag) {
+    return new Response(null, { status: 304 })
+  }
+
+  if (request.method === 'HEAD') {
+    const obj = await env.CARPARK.head(`${dataCid}/${dataCid}.car`)
+    if (!obj) throw new HttpError('CAR not found', { status: 404 })
+    return new Response(undefined, {
+      headers: {
+        'Accept-Ranges': 'bytes',
+        'Content-Length': obj.size.toString(),
+        Etag: etag
+      }
+    })
+  }
+
+  /** @type {import('../lib/http').Range|undefined} */
+  let range
+  if (request.headers.has('range')) {
+    try {
+      range = Http.parseRange(request.headers.get('range') ?? '')
+    } catch (err) {
+      throw new HttpError('invalid range', { status: 400, cause: err })
+    }
+  }
+
+  const obj = await env.CARPARK.get(`${dataCid}/${dataCid}.car`, { range })
```

Comment:
Could be cool to try this with roundabout. I think it should work and let us redirect to presigned url. 

Probably good to keep as is and track issue to attempt cost optimisation :)

Reply:
Hmm, yes but we probably don't want to give out those URLs to anyone via the gateway. Someone will try to use/abuse and they also expire...

Reply:
So the usage would be the same as here? it would be a redirect that response will be read right away right? So, expiration would not be a problem, and if we perform request to `roundabout` within the worker and redirect to presigned URL then user will not even know how URL was created.

Use/abuse would be the same as via freeway no. So, rate limits is really the solution we seem to have to avoid abuse